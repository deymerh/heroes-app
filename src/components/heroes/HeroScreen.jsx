import React, { useMemo } from 'react';
import { Redirect, useHistory, useParams } from 'react-router-dom';
import { getHeroesById } from '../../selectors/getHeroesById';

export const HeroScreen = () => {
  let { heroesId } = useParams();
  let history = useHistory();
  const hero = useMemo(() => getHeroesById(heroesId), [heroesId]);
  if (!hero) {
    return <Redirect to="/" />
  }
  const {
    superhero,
    publisher,
    alter_ego,
    first_appearance,
    characters,
  } = hero;
  const handleReturn = () => {
    if (history.length <= 2) {
      history.push('/');
    } else {
      history.goBack();
    }
  }
  return (
    <div className="row mt-5">
      <div className="col-4">
        <img src={`../assets/heroes/${heroesId}.jpg`} alt={superhero} className="img-thumbnail animate__animated animate__fadeInLeft" />
      </div>
      <div className="col-8">
        <h3>{superhero}</h3>
        <ul className="list-group list-group-flush">
          <li className="list-group-item"> <b>Alter ego: </b>{alter_ego} </li>
          <li className="list-group-item"> <b>Publisher: </b>{publisher} </li>
          <li className="list-group-item"> <b>First appearance: </b>{first_appearance} </li>
        </ul>
        <h3> Characters</h3>
        <p>{characters}</p>
        <button className="btn btn-outline-info" onClick={handleReturn}>
          Return
        </button>
      </div>
    </div>
  )
};
