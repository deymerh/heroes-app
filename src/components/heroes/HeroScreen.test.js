import { mount } from 'enzyme';
import { MemoryRouter, Route, Router } from 'react-router-dom';
import { HeroScreen } from './HeroScreen';

describe('Pruebas en el componente <HeroScreen />', () => {
  
  const historyMock = {
    length: 10,
    push: jest.fn(),
    goBack: jest.fn()
  }
  test('Debe retornar el componente `Redirec` si no hay argumentos en el URL', () => {
    const wrapper = mount(
      <MemoryRouter  initialEntries={['/hero']}>
        <HeroScreen history={historyMock} />
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('Redirect').exists()).toBeTruthy();
  });
  
  test('Debe mostrar si el parametro existe y se encuentra', () => {
    const wrapper = mount(
      <MemoryRouter  initialEntries={['/hero/marvel-spider']}>
        <Route path="/hero/:heroesId" component={HeroScreen} />
      </MemoryRouter>
    );
    expect(wrapper.find('.row').exists()).toBe(true);
  });
  
  test('Debe regresar a la pantalla anterior con el PUSH', () => {
    const history = {
      length: 1,
      push: jest.fn(),
      goBack: jest.fn(),
      listen: jest.fn(),
      location:{pathname:'/hero/marvel-spider'},
    }
    const wrapper = mount(
      <MemoryRouter>
        <Router history={history}>
          <Route 
            path="/hero/:heroesId"
            component={HeroScreen}
            />
        </Router>
      </MemoryRouter>
    );
    wrapper.find('button').prop('onClick')();
    expect(history.push).toHaveBeenCalled()
    expect(history.goBack).not.toHaveBeenCalled()
  });
  
  test('Debe regresar a la pantalla anterior GOBACK', () => {
      const history = {
        length: 3,
        push: jest.fn(),
        goBack: jest.fn(),
        listen: jest.fn(),
        location:{pathname:'/hero/marvel-spider'},
      }
      const wrapper = mount(
        <MemoryRouter>
          <Router history={history}>
            <Route 
              path="/hero/:heroesId"
              component={HeroScreen}
              />
          </Router>
        </MemoryRouter>
      );
      wrapper.find('button').prop('onClick')();
      expect(history.goBack).toHaveBeenCalled()
      expect(history.push).not.toHaveBeenCalled()
  });
  
});     
