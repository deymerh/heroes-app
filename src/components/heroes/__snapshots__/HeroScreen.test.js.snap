// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <HeroScreen /> Debe retornar el componente \`Redirec\` si no hay argumentos en el URL 1`] = `""`;
